clients/vault/src/oracle/collector/proof_builder.rs

```diff
@@ -1,160 +1,261 @@
-use crate::oracle::{ActorMessage, ScpArchiveStorage};
-use stellar_relay::sdk::{
-	compound_types::{LimitedVarArray, UnlimitedVarArray, XdrArchive},
-	types::{ScpEnvelope, ScpHistoryEntry, TransactionSet},
-	TransactionEnvelope, XdrCodec,
+use crate::oracle::ScpArchiveStorage;
+use parking_lot::RwLock;
+use std::{convert::TryInto, sync::Arc};
+use stellar_relay::{
+	sdk::{
+		compound_types::{UnlimitedVarArray, XdrArchive},
+		types::{ScpEnvelope, ScpHistoryEntry, StellarMessage, TransactionSet},
+		XdrCodec,
+	},
+	StellarOverlayConnection,
 };
 
 use crate::oracle::{
 	constants::{get_min_externalized_messages, MAX_SLOT_TO_REMEMBER},
-	traits::FileHandler,
-	EnvelopesFileHandler, ScpMessageCollector, Slot, TxHash, TxSetsFileHandler,
+	types::EnvelopesMap,
+	ScpMessageCollector, Slot,
 };
 
-use std::convert::TryInto;
-
-/// Determines whether the data retrieved is from the current map or from a file.
-type DataFromFile<T> = (T, bool);
-
 /// The Proof of Transactions that needed to be processed
+#[derive(Debug, Eq, PartialEq)]
 pub struct Proof {
-	tx_env: TransactionEnvelope,
+	slot: Slot,
 	envelopes: UnlimitedVarArray<ScpEnvelope>,
 	tx_set: TransactionSet,
 }
 
 impl Proof {
 	/// Encodes these Stellar structures to make it easier to send as extrinsic.
-	pub fn encode(&self) -> (String, String, String) {
-		let tx_env_xdr = self.tx_env.to_xdr();
-		let tx_env_encoded = base64::encode(tx_env_xdr);
-
+	pub fn encode(&self) -> (String, String) {
 		let envelopes_xdr = self.envelopes.to_xdr();
 		let envelopes_encoded = base64::encode(envelopes_xdr);
 
 		let tx_set_xdr = self.tx_set.to_xdr();
 		let tx_set_encoded = base64::encode(tx_set_xdr);
 
-		(tx_env_encoded, envelopes_encoded, tx_set_encoded)
+		(envelopes_encoded, tx_set_encoded)
+	}
+
+	pub fn slot(&self) -> Slot {
+		self.slot
+	}
+
+	pub fn envelopes(&self) -> &Vec<ScpEnvelope> {
+		self.envelopes.get_vec()
+	}
+
+	pub fn tx_set(&self) -> &TransactionSet {
+		&self.tx_set
 	}
 }
 
+#[derive(Debug, Eq, PartialEq)]
 pub enum ProofStatus {
 	Proof(Proof),
-	LackingEnvelopes(Slot),
-	NoTxSetFound(Slot),
+	/// The available ScpEnvelopes are not enough to create a proof, and the slot is too far back
+	/// to fetch more.
+	LackingEnvelopes,
+	/// More envelopes have been requested from the Stellar Node, so user needs to wait.
+	WaitForMoreEnvelopes,
+	/// No ScpEnvelopes found and the slot is too far back.
+	NoEnvelopesFound,
+	/// Envelopes is fetched again, and user just needs to wait for it.
+	WaitForEnvelopes,
+	/// TxSet is not found and the slot is too far back.
```

Comment:
Whats the difference between `WaitForEnvelopes` and `WaitForMoreEnvelopes`?

Reply:
removed extra variants